# Семинар 1: Введение в Python

## Организационные вопросы

### Внеаудиторное общение

**Преподаватель:** Бурова Маргарита Борисовна

* почта: mbburova@gmail.com
* 89264931203
* страница курса: http://wiki.cs.hse.ru/Python_для_ОП_%22Психология%22_2021/2022
* чаты групп в телеграме

### Формула оценки

$ 0.2 \cdot ДЗ 1 + 0.2 \cdot ДЗ 2 + 0.1 \cdot Онлайн + 0.25 \cdot СР + 0.25 \cdot КР $

* **ДЗ1 и ДЗ2** - Индивидуальные ДЗ
> Необходимо выполнить 10 критериев, за каждый из критериев ставится 0, 0.5 или 1. Оценка не округляется
* **Онлайн** - Оценка, полученная за онлайн-курс
> Процент выполненных заданий, деленный на 10. Оценка не округляется
* **СР** - Среднее арифметическое по четырем самостоятельным, которые будут проведены на семинарах
> Самостоятельные в формате компьютерного тестирования, примерно 20-25 минут
* **КР** - Контрольная работа в формате, соответствующем спецификации НОК, состоящий из тестовых вопросов и задач

### Где мы будем работать?

Jupyter Notebook - интерактивная среда для запуска программного кода в браузере. Удобный инструмент для анализа данных, который используется многими специалистами по data science.

![](https://media.giphy.com/media/5Zesu5VPNGJlm/giphy.gif)

###  Меньше текста, больше кода

**Напутствие**:
- *главное, не бояться пробовать писать код*
- *каким бы неправильным он ни казался*
- *скорее всего, ничего такого, что не починится, в рамках Jupyter Notebook вы сделать не сможете* 

**Краткая инструкция на все случаи жизни**:
- Как создать новую ячейку?

> Просто нажать кнопку `+` (*insert cell below*).

- Как запустить код в ячейке?

> Выделить ячейку и нажать кнопку `Run` или `Ctrl + Enter` *(см. Help - Keyboard Shortcuts)*.

- Как понять, что код успешно выполнен?

> Под ячейкой не появились сообщения об ошибках + поле `In []` слева от ячейки содержит число, а не символ `*`.

- Что делать, если код не хочет успешно выполняться, поле слева выглядит вот так `In [*]` и вообще всё зависло аааа помогите
<img src="https://pp.userapi.com/c543103/v543103112/326eb/UnK2PZuGroQ.jpg" alt="Drawing" style="width: 200px;"/>

> Для прерывания выполнения кода можно попробовать нажать кнопку `■` (*interrupt the kernel*) в панели наверху, либо перезапустить весь проект, нажав соседнюю кнопку (*restart the kernel*).


In [2]:
2+10

12

## План на сегодня

* Функции и вывод данных
* Основные типы данных : число, текст, логические переменные
* Действия с основными типами данных
* Переменные

## Функции и вывод на экран

В дальнейшем вы часто будете использовать _функции_ из Python или из специализированных Python-модулей.

_Функция_ -- это какой-то скрытый от вас программный код, который можно применять для решения конкретной задачи.

Например, познакомимся с функцией `print`:
- нужна для того, чтобы выводить на экран текст / числа / что угодно ещё
- формат вызова: `print(<первое, что хотите вывести на экран>, <второе, что хотите вывести на экран>, ...)`

То, что перечисляется внутри скобок и разделяется запятыми, называется *аргументами* или _входами_ функции

<img src="https://i.ibb.co/12sSpRd/Untitled-Frame-1-11.jpg" alt="Drawing" style="width: 800px;"/>

In [3]:
print(67+87)
print(42+146)

154
188


In [4]:
print('Hello world!')

Hello world!


Попробуем вывести результат сложения чисел 42 и 146:

In [23]:
print(42+146)
# всё, что пишется за знаком #, воспринимается как комментарий и не выполняется в рамках программы

188


## Типы данных: Целые числа - INT (INTEGER)

Python сам понимает, какой тип у данных. Однако, можно так же и запросить у него данную информацию

In [1]:
print(type(5))

<class 'int'>


Разумеется, в python представлены все возможные арифметические операции:

In [25]:
# сложение
print(4 + 2)
# вычитание
print(4 - 2)

6
2


In [26]:
# умножение
print(4 * 2)
# деление
print(4 / 2)
# взятие остатка от деления
print(3 % 2)

8
2.0
1


In [27]:
# возведение в степень
print(3 ** 2)
# взятие корня aka возведение в дробную степень
print(3 ** (1 / 2))

9
1.7320508075688772


Заметим, что здесь у нас в ответах получились не только целые числа, но также дроби.

## Типы данных: Вещественные числа - FLOAT

In [28]:
type(1.5)

float

На что обратить внимание:
- числа с десятичной частью, например, $1.43$, в отличие от Excel, в Python пишутся не через запятую `,`, а через точку `.`.

Почувствуйте разницу:

In [29]:
print(2,6 / 2)  # печатает два числа
print(2, 6 / 2)  # и так тоже два числа, пробел в данном случае неважен
print(2.6 / 2)  # печатает одно число

2 3.0
2 3.0
1.3


Также не стоит путать целочисленное и вещественное деление:

In [30]:
print(9 / 2)  # печатает ожидаемый дробный результат
print(9 // 2)  # целочисленное деление
print(9 % 2)  # остаток от (целочисленного) отделения

4.5
4
1


Иногда результаты могут быть довольно неожиданные. Может показаться, что python совершенно не умеет считать:

In [31]:
print(0.1+0.2)

0.30000000000000004


Еще результат может выводиться в довольно странном формате:

In [32]:
1.543**1000

2.3223413936165768e+188

Такая запись расшифровывается следующим образом:

$ n \cdot e+188 = n \cdot 10^{188} $

## Типы данных: Строки (текст) - STR (STRING)

С помощью `print` можно печатать не только числа. Например, текст:
- текстовая строка в питоне -- это любое выражение, заключённое в двойные `"` или одинарные `'` кавычки
- с точки зрения пользователя ничем не отличается от привычного текста

Например:

In [33]:
print('Моё имя:')

Моё имя:


In [34]:
type('Маргарита')

str

In [5]:
print('Первая'+'пара')

Перваяпара


In [8]:
print('У меня '+ 'проблема'*20)

У меня проблемапроблемапроблемапроблемапроблемапроблемапроблемапроблемапроблемапроблемапроблемапроблемапроблемапроблемапроблемапроблемапроблемапроблемапроблемапроблема


Ещё строки можно
- складывать
- умножать на число
- см. встроенные методы класса `str`

**lifehack**: *если написать `str.` и нажать клавишу `Tab`, появится список доступных для строк методов*.

In [35]:
print('how')
print('how' + ' ' + 'are')
print('how' * 3)
print(' are' * 2 + 'you?')

how
how are
howhowhow
 are areyou?


На что обратить внимание:
- выражение `1` это число
- выражение `'1'` это строка

Почувствуйте разницу:

In [36]:
print(1 + 2)  # печатает число
print('1' + '2')  # печатает результат склеивания строк
print(str(1) + str(2))  # функция str преобразует аргумент в строку

3
12
12


## Типы данных: Логические переменные - boolean

In [9]:
print(type(False))

<class 'bool'>


In [38]:
print(type('True'))

<class 'str'>


In [39]:
print(34 == 30+4)

True


In [40]:
print(35 == 30-5)

False


## Ошибки в Python и что с ними делать

Иногда написанный код не работает. Так бывает. Это не страшно.

Например, я хочу вывести на экран текст следующего вида:
```
День недели: суббота.
Моё имя: Маргарита
Мой возраст: 28
```

Я написала свой код в ячейке ниже.

Выполните следующую ячейку (казалось бы, что может пойти не так?):

In [11]:
print('День недели: суббота.')
print('Моё имя: ' + 'Маргарита')
print('Мой возраст: ' + '28')

День недели: суббота.
Моё имя: Маргарита
Мой возраст: 28


## Переменные

В Python мы можем записывать в переменные какие-либо значения (или наборы значений):

In [42]:
number = 99
my_name = 'Маргарита'

In [43]:
print(number)
print(my_name)

99
Маргарита


Если мы будем реализовывать какие-то действия с переменными, то они будут осуществляться со значениями, которые записаны в эти переменные:

In [44]:
print(number*my_name)

МаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргарита
